File: src/data/4o/Sorted_GCD_Pair_Queries/Sorted_GCD_Pair_Queries.java
import java.util.*;

class Solution {
    public int[] gcdValues(int[] nums, long[] queries) {
        List<Integer> gcdPairs = new ArrayList<>();
        int n = nums.length;

        // Calculate GCD for all possible pairs and store them in gcdPairs
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                gcdPairs.add(gcd(nums[i], nums[j]));
            }
        }

        // Sort gcdPairs in ascending order
        Collections.sort(gcdPairs);

        // Answer each query
        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            if (queries[i] < gcdPairs.size()) {
                answer[i] = gcdPairs.get((int) queries[i]);
            } else {
                answer[i] = -1; // If the query is out of bounds
            }
        }

        return answer;
    }

    // Helper function to calculate GCD
    private int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }
}


public class Sorted_GCD_Pair_Queries {
    public static void runTests(Solution solution) {
        // Test case 1
        int[] nums = {2, 3};
        long[] queries = {0};
        int[] expectedResult = {1};  // Only one pair, GCD(2, 3) = 1
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 1 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 2
        nums = new int[]{2, 3, 4};
        queries = new long[]{0, 2, 2};
        expectedResult = new int[]{1, 2, 2};  // GCD pairs sorted: [2, 2, 1]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 2 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 3
        nums = new int[]{4, 8, 12};
        queries = new long[]{0, 1, 2};
        expectedResult = new int[]{4, 4, 4};  // GCD pairs sorted: [4, 4, 4]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 3 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 4
        nums = new int[]{2, 3, 5, 7};
        queries = new long[]{0, 1, 2, 3, 4, 5};
        expectedResult = new int[]{1, 1, 1, 1, 1, 1};  // GCDs are all 1 for prime pairs
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 4 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 5
        nums = new int[]{4, 4, 2, 1};
        queries = new long[]{5, 3, 1, 0};
        expectedResult = new int[]{4, 2, 1, 1};  // GCD pairs sorted: [4, 4, 2, 2, 2, 1]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 5 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 6
        nums = new int[]{100000, 100000};
        queries = new long[]{0};
        expectedResult = new int[]{100000};  // Only one GCD of the pair: [100000]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 6 failed. Expected " + java.util.Arrays.toString(expectedResult);
    }
    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 20000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/4o/Sorted_GCD_Pair_Queries/Sorted_GCD_Pair_Queries.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> gcdValues(vector<int>& nums, vector<long long>& queries) {
        int n = nums.size();
        vector<int> gcdPairs;

        // Compute the GCD of all possible pairs
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                gcdPairs.push_back(__gcd(nums[i], nums[j]));
            }
        }

        // Sort the GCD values
        sort(gcdPairs.begin(), gcdPairs.end());

        // Answer queries based on the sorted gcdPairs array
        vector<int> answer;
        for (long long q : queries) {
            if (q < gcdPairs.size()) {
                answer.push_back(gcdPairs[q]);
            } else {
                answer.push_back(-1); // Return -1 if the query index is out of bounds
            }
        }

        return answer;
    }
};

//  Test cases for the C++ solution
#include <cassert>

void testSolution(Solution& solution) {

    // Test case 1: Simple case with two numbers
    std::vector<int> nums = {2, 3};
    std::vector<long long> queries = {0};
    std::vector<int> expected_result = {1};  // Only one pair, GCD(2, 3) = 1
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 1 failed. Expected {1}.");

    // Test case 2: Multiple pairs with diverse GCDs
    nums = {2, 3, 4};
    queries = {0, 2, 2};
    expected_result = {1, 2, 2};  // GCD pairs sorted: [2, 2, 1]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 2 failed. Expected {1, 2, 2}.");

    // Test case 3: All numbers are multiples of a common number
    nums = {4, 8, 12};
    queries = {0, 1, 2};
    expected_result = {4, 4, 4};  // GCD pairs sorted: [4, 4, 4]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 3 failed. Expected {4, 4, 4}.");

    // Test case 4: GCD results in multiple distinct values
    nums = {2, 3, 5, 7};
    queries = {0, 1, 2, 3, 4, 5};
    expected_result = {1, 1, 1, 1, 1, 1};  // GCDs are all 1 for prime pairs
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 4 failed. Expected {1, 1, 1, 1, 1, 1}.");

    // Test case 5: GCD values with repeated elements in nums
    nums = {4, 4, 2, 1};
    queries = {5, 3, 1, 0};
    expected_result = {4, 2, 1, 1};  // GCD pairs sorted: [4, 4, 2, 2, 2, 1]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 5 failed. Expected {4, 2, 1, 1}.");

    // Test case 6: Large number with single pair
    nums = {100000, 100000};
    queries = {0};
    expected_result = {100000};  // Only one GCD of the pair: [100000]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 6 failed. Expected {100000}.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 1280; ++i) {
        testSolution(solution);  
    }
    return 0;
}
